pkg/controllers/workapplier/apply.go

```diff
@@ -0,0 +1,539 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package workapplier
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils/resource"
+)
+
+var builtInScheme = runtime.NewScheme()
+
+func init() {
+	// This is a trick that allows Fleet to check if a resource is a K8s built-in one.
+	_ = clientgoscheme.AddToScheme(builtInScheme)
+}
+
+// applyInDryRunMode
+func (r *Reconciler) applyInDryRunMode(
+	ctx context.Context,
+	gvr *schema.GroupVersionResource,
+	manifestObj, inMemberClusterObj *unstructured.Unstructured,
+) (*unstructured.Unstructured, error) {
+	// In this method, Fleet will always use forced server-side apply
+	// w/o optimistic lock for diff calculation.
+	//
+	// This is OK as partial comparison concerns only fields that are currently present
+	// in the manifest object, and Fleet will clear out system managed and read-only fields
+	// before the comparison.
+	//
+	// Note that full comparison can be carried out directly without involving the apply op.
+	return r.serverSideApply(ctx, gvr, manifestObj, inMemberClusterObj, true, false, true)
+}
+
+func (r *Reconciler) apply(
```

Comment:
we should try to differentiate the type of errors
1. userError
2. none-retriable error
3. retriable error

Reply:
This is very important when we get to upgrade so that we know if it's something we can continue or give up

Reply:
Done; the categorization is done slightly different from other controllers: it's not passed back as there are internal only information that should not be exposed to the end user in the K8s condition; some errors (e.g., drift/diff found) are not registered as user errors as they are part of the flow.